sql/information_schema/columns_table.go

```diff
@@ -185,45 +186,41 @@ func columnsRowIter(ctx *sql.Context, catalog sql.Catalog, allColsWithDefaultVal
 // getRowFromColumn returns a single row for given column. The arguments passed are used to define all row values.
 // These include the current ordinal position, so this column will get the next position number, sql.Column object,
 // database name, table name, column key and column privileges information through privileges set for the table.
-func getRowFromColumn(ctx *sql.Context, curOrdPos int, col *sql.Column, dbName, tblName, columnKey string, privSetTbl sql.PrivilegeSetTable, privSetMap map[string]struct{}) sql.Row {
+func getRowFromColumn(ctx *sql.Context, curOrdPos int, col *sql.Column, dbCollation sql.CollationID, dbName, tblName, columnKey string, privSetTbl sql.PrivilegeSetTable, privSetMap map[string]struct{}) sql.Row {
 	var (
 		charName          interface{}
 		collName          interface{}
 		charMaxLen        interface{}
 		charOctetLen      interface{}
 		nullable          = "NO"
 		ordinalPos        = uint32(curOrdPos + 1)
-		colType           = strings.Split(col.Type.String(), " COLLATE")[0]
-		dataType          = colType
 		datetimePrecision interface{}
 		srsId             interface{}
 	)
 
+	colType, dataType := getDtdIdAndDataType(col.Type)
+
 	if col.Nullable {
 		nullable = "YES"
 	}
 
 	if types.IsText(col.Type) {
 		if types.IsTextOnly(col.Type) {
-			charName = sql.Collation_Default.CharacterSet().String()
-			collName = sql.Collation_Default.String()
+			charName = dbCollation.CharacterSet().String()
+			collName = dbCollation.String()
 		}
 
 		if st, ok := col.Type.(types.StringType); ok {
 			charMaxLen = st.MaxCharacterLength()
 			charOctetLen = st.MaxByteLength()
 		}
 	} else if types.IsEnum(col.Type) || types.IsSet(col.Type) {
-		charName = sql.Collation_Default.CharacterSet().String()
-		collName = sql.Collation_Default.String()
+		charName = dbCollation.CharacterSet().String()
+		collName = dbCollation.String()
 		charOctetLen = int64(col.Type.MaxTextResponseByteLength())
-		charMaxLen = int64(col.Type.MaxTextResponseByteLength()) / sql.Collation_Default.CharacterSet().MaxLength()
+		charMaxLen = int64(col.Type.MaxTextResponseByteLength()) / dbCollation.CharacterSet().MaxLength()
```

Comment:
After using that interface I mentioned above, you can then special case `Enum` and `Set` to set the lengths here since the interface already takes care of the charset and collation